Pass additional flags to subpath mount to avoid flakes in certain conditions

mauriciopoppe opened the pull request:
#### What type of PR is this?

/kind bug
/sig storage

#### What this PR does / why we need it:

Pass additional flags to subpath mount to avoid flakes in certain conditions, supersedes #104254, the new strategy is to create another mount implementation only for subpath usage.

- Copy the git history of mount_linux.go, mount_windows.go, mount.go to the subpath directory
- Create a new MountInterface in subpath that includes the k8s.io/utils/mount interface, in addition it declares a new method `MountSensitiveWithMountFlags` that will be used only for subpath mount purposes
- Make the subpath interface use the new MountInterface instead of k8s.io/utils/mount
- When an instance of subpather is created instead of using the k8s.io/util/mount instance we use an instance of the new MountInterface
- Finally use `MountSensitiveWithMountFlags` in subpath_linux.go:doBindSubpath()

Also I created a GCE cluster and verified it in the kubelet logs

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Pass additional flags to subpath mount to avoid flakes in certain conditions
```

/cc @msau42 @jingxu97 @jsafrane

mauriciopoppe commented:
/retest

mauriciopoppe committed:
Duplicate vendor/k8s.io/utils/mount/* history in pkg/volume/util/subpath/subpath_* history.

msau42 commented on pkg/volume/util/subpath/subpath_mount_linux.go:
```diff
@@ -0,0 +1,278 @@
+// +build linux
```
We synced offline and potentially found a way to extend the mount interface without needing to copy the original implementations.

mauriciopoppe committed:
New subpath Mounter implements mount.Interface, update kubelet startup

- The kubelet startup code will create a new Mounter instance
  that wraps the existing k8s.io/utils/mount instance
- Provided implementations for all the platforms

mauriciopoppe committed:
Add MountSensitiveWithFlags to the subpath Mounter interface

- Also adding an implementation for all the platforms

mauriciopoppe committed:
Duplicate vendor/k8s.io/utils/mount/fake_mounter.go history in pkg/volume/util/subpath/subpath_fake_mounter.go history.

mauriciopoppe committed:
Create a FakeMounter for the subpath that wraps the mountutils FakeMounter

mauriciopoppe committed:
Pass additional flags to subpath mount to avoid flakes in certain conditions

mauriciopoppe committed:
Add subpath mount files to BUILD

jingxu97 commented:
some bazel test failed?

mauriciopoppe commented:
@jingxu97 yes, I fixed it in 2bfa2d7

@msau42 @jsafrane @jingxu97 this is ready for review, thanks for the hint @msau42!

msau42 commented:
/lgtm
/approve

mauriciopoppe commented:
hey @Random-Liu, I'd really appreciate a review on this PR, in summary the kubelet subpather is receiving a new mounter that wraps k8s.io/utils/mount

mauriciopoppe commented:
/cc @Random-Liu

yujuhong commented:
/approve

for the cmd/kubelet change.

mauriciopoppe commented:
hey @kubernetes/release-managers, could you please review this PR? thanks in advance!

mauriciopoppe commented:
hey @cpanato, I saw that you helped approving https://github.com/kubernetes/kubernetes/pull/104253, this PR is similar but for 1.19, do you mind taking a look please? thanks

jsturtevant commented:
/sig windows

cpanato commented:
looks like this was not cherry-picked using the process

cpanato commented:
/triage accepted

from parent PR:
/priority important-soon

cpanato reviewed (approved):
from RelEng:
/lgtm
